plugins/StereoControl/StereoControl.cpp

```diff
@@ -0,0 +1,443 @@
+						if (readLoc < m_haasBufSize - 1)
+						{
+							haasResult = linearInterpolate(m_haasBuf[i][floor(readLoc)], m_haasBuf[i][ceil(readLoc)], readLocFrac);
+						}
+						else// For when the interpolation wraps around to the beginning of the buffer
+						{
+							haasResult = linearInterpolate(m_haasBuf[i][m_haasBufSize - 1], m_haasBuf[i][0], readLocFrac);
+						}
+					}
+
+					// Increment delay ring buffer index
+					++m_haasIndex[i];
+					if (m_haasIndex[i] >= m_haasBufSize)
+					{
+						m_haasIndex[i] -= m_haasBufSize;
+					}
+					// Send new value to delay line
+					m_haasBuf[i][m_haasIndex[i]] = s[i];
+				}
+
+				if (pan != 0)
+				{
+					const float minGain = linearInterpolate(1.f, 0.15f, abs(pan));
+					if (pan >= 0)
+					{
+						const float lGain = ((1.f - pan) + minGain) / (1.f + minGain);
+						s[0] = haasResult * lGain;
+					}
+					else
+					{
+						const float rGain = ((1.f + pan) + minGain) / (1.f + minGain);
+						s[1] = haasResult * rGain;
+					}
+				}
+
+				const int lowChnl = (pan <= 0);
+				const int highChnl = !lowChnl;
+
+				float lp;
+				float hp;
+
+				// Filter parameters are semi-arbitrary, but are vaguely based on some random graphs I saw during research
+				float temp = -abs(pan) + 1.f;
+				float filtFreq = 2000.f + 8000.f * temp * temp;
+				float hpGain = abs(pan) < 0.5 ? 1.f - abs(pan) * 2.f : 0.f;
+				hpGain = linearInterpolate(1.f, hpGain, panSpectral);
+				multimodeFilter(s[lowChnl], multimodeCoeff(filtFreq), lp, hp, m_haasSpectralPanFilter[lowChnl]);
+
+				s[lowChnl] = lp + hp * hpGain;
+
+				filtFreq = 2000.f;
+				hpGain = abs(pan) + 1.f;
+				hpGain = linearInterpolate(1.f, hpGain, panSpectral);
+				multimodeFilter(s[highChnl], multimodeCoeff(filtFreq), lp, hp, m_haasSpectralPanFilter[highChnl]);
+
+				s[highChnl] = lp + hp * hpGain;
+			}
+		}
+
+		if (dc)
+		{
+			for (int i = 0; i < 2; ++i)
+			{
+				m_dcInfo[i] = m_dcInfo[i] * 0.999f + s[i] * 0.001f;
+				s[i] -= m_dcInfo[i];
+			}
+		}
+
+		if (mute)
+		{
+			s[0] = 0;
+			s[1] = 0;
+		}
+
+		s[0] *= gain;
+		s[1] *= gain;
+
+		buf[f][0] = d * buf[f][0] + w * s[0];
+		buf[f][1] = d * buf[f][1] + w * s[1];
+
+		outSum += buf[f][0] * buf[f][0] + buf[f][1] * buf[f][1];
+	}
+
+	checkGate(outSum / frames);
+
+	return isRunning();
+}
+
+
+sample_t StereoControlEffect::calcAllpassFilter(sample_t inSamp, sample_rate_t Fs, int filtNum, int channel, float a, float b)
+{
+	float filterOutput = a * (inSamp - m_filtY[filtNum][channel][1]) +
+		b * (m_filtX[filtNum][channel][0] - m_filtY[filtNum][channel][0]) +
+		m_filtX[filtNum][channel][1];
+
+	m_filtX[filtNum][channel][1] = m_filtX[filtNum][channel][0];
+	m_filtX[filtNum][channel][0] = inSamp;
+	m_filtY[filtNum][channel][1] = m_filtY[filtNum][channel][0];
+	m_filtY[filtNum][channel][0] = filterOutput;
+
+	return filterOutput;
+}
+
+
+void StereoControlEffect::multimodeFilter(sample_t in, float g, sample_t &lp, sample_t &hp, sample_t &filtBuf)
```

Comment:
Rather than passing different member variables of `StereoControlEffect` into here as references, why not make them members of a new `MultimodeFilter` class? Also, you can return multiple values using something like `std::pair<sample_t, sample_t>` or `std::array<sample_t, 2>`, meaning you don't have to use out parameters. And `multimodeCoeff` is only and always used as a parameter to this function, so could be computed inside of it.

Code like this:
```cpp
float lp;
float hp;
float hp2;
multimodeFilter(allpassValue, multimodeCoeff(stereoizerHP), lp, hp, m_stereoizerFilter[0][i]);
multimodeFilter(hp, multimodeCoeff(stereoizerLP), lp, hp2, m_stereoizerFilter[1][i]);
s[i] += lp * qMin(stereoizer, 1.f);
```
could look like this:
```cpp
const auto stereoHigh = m_stereoizerFilter[0][i].update(allpassValue, stereoizerHP);
const auto stereoLow = m_stereoizerFilter[1][i].update(stereoHigh[1], stereoizerLP);
s[i] += stereoLow[0] * qMin(stereoizer, 1.f);
```

This takes fewer lines since the output parameters don't need declaring, is clearer as to what is input and what is output, doesn't require a separate call to pre-process the frequency, and better associates the member variables with the code that uses them.